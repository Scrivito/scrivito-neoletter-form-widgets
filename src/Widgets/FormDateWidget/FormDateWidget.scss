.scrivito-neoletter-form-widgets .form-date {
  .date-title {
    p {
      display: inline;
    }
  }
}
