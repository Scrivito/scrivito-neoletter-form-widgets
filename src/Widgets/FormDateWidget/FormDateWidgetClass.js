import * as Scrivito from "scrivito";

export const FormDateWidget = Scrivito.provideWidgetClass("FormDateWidget", {
  attributes: {
    title: "string",
    helpText: "html",
    required: "boolean",
    customFieldName: "string",
    dateType: [
      "enum",
      {
        values: ["date", "datetime-local"],
      },
    ],
  },
});
