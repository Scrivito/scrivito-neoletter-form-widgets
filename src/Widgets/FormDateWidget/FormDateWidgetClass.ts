import * as Scrivito from "scrivito";

export const FormDateWidget = Scrivito.provideWidgetClass("FormDateWidget", {
  attributes: {
    title: "html",
    helpText: "html",
    required: "boolean",
    customFieldName: "string",
    dateType: [
      "enum",
      {
        values: ["date", "datetime-local"]
      }
    ]
  }
});
