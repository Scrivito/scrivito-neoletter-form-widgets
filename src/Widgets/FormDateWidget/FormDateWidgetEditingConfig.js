import * as Scrivito from "scrivito";
import { customFieldNameValidation } from "../FormContainerWidget/utils/validations/customFieldNameValidation";
import { insideFormContainerValidation } from "../FormContainerWidget/utils/validations/insideFormContainerValidation";
import formDateWidgetIcon from "../../assets/images/form_widget_date.svg";

Scrivito.provideEditingConfig("FormDateWidget", {
  title: "Form Date",
  thumbnail: formDateWidgetIcon,
  attributes: {
    title: { title: "Label" },
    dateType: {
      title: "Format",
      values: [
        { value: "date", title: "Date" },
        { value: "datetime-local", title: "Date and time" },
      ],
    },
    customFieldName: { title: "Field name" },
    helpText: { title: "Help text" },
    required: { title: "Mandatory" },
  },
  properties: ["title", "dateType", "customFieldName", "required", "helpText"],
  initialContent: {
    title: "Please enter a date",
    customFieldName: "custom_",
    dateType: "date",
  },
  validations: [insideFormContainerValidation, customFieldNameValidation],
});
