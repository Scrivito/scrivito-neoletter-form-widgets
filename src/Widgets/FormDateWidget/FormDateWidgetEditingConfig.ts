import * as Scrivito from "scrivito";
import { customFieldNameValidation } from "../FormStepContainerWidget/utils/validations/customFieldNameValidation";
import { insideFormContainerValidation } from "../FormStepContainerWidget/utils/validations/insideFormContainerValidation";
import formDateWidgetIcon from "../../assets/images/form_widget_date.svg";

Scrivito.provideEditingConfig("FormDateWidget", {
  title: "Neoletter Form Date",
  thumbnail: formDateWidgetIcon,
  attributes: {
    title: { title: "Label" },
    dateType: {
      title: "Format",
      values: [
        { value: "date", title: "Date" },
        { value: "datetime-local", title: "Date and time" }
      ]
    },
    customFieldName: { title: "Field name" },
    helpText: { title: "Help text" },
    required: { title: "Mandatory" }
  },
  properties: ["title", "dateType", "customFieldName", "required", "helpText"],
  initialContent: {
    title: "Please enter a date",
    customFieldName: "custom_",
    dateType: "date"
  },
  validations: [insideFormContainerValidation, customFieldNameValidation]
});
