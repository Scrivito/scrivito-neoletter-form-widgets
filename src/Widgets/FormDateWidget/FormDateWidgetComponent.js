import * as React from "react";
import * as Scrivito from "scrivito";
import { getFieldName } from "../FormContainerWidget/utils/getFieldName";
import { HelpText } from "../FormContainerWidget/components/HelpTextComponent";
import { Mandatory } from "../FormContainerWidget/components/MandatoryComponent";
import "./FormDateWidget.scss";

Scrivito.provideComponent("FormDateWidget", ({ widget }) => {
  const [value, setValue] = React.useState("");
  const onChangeValue = (e) => {
    const d = new Date(e.target.value);
    setValue(d.toISOString());
  };
  return (
    <div className="form-date mb-3">
      <div className="date-title">
        <span className="text-super">{widget.get("title")}</span>
        {widget.get("required") && <Mandatory />}
        {widget.get("helpText") && <HelpText widget={widget} />}
      </div>

      <input
        onChange={onChangeValue}
        className="datepicker"
        type={widget.get("dateType")}
        required={widget.get("required")}
      ></input>
      <input type="hidden" name={getFieldName(widget)} value={value}></input>
    </div>
  );
});
