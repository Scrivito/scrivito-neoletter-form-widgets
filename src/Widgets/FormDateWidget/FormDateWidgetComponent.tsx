import * as React from "react";
import * as Scrivito from "scrivito";
import { getFieldName } from "../FormStepContainerWidget/utils/getFieldName";
import { HelpText } from "../FormStepContainerWidget/components/HelpTextComponent";
import { Mandatory } from "../FormStepContainerWidget/components/MandatoryComponent";
import { FormDateWidget } from "./FormDateWidgetClass";
import "./FormDateWidget.scss";

Scrivito.provideComponent(FormDateWidget, ({ widget }) => {
  const [value, setValue] = React.useState("");
  const onChangeValue = (e: React.BaseSyntheticEvent) => {
    const d = new Date(e.target.value);
    setValue(d.toISOString());
  };
  return (
    <div className="form-date mb-3">
      <div className="date-title">
        <span className="text-super">{widget.get("title")}</span>
        {widget.get("required") && <Mandatory />}
        {widget.get("helpText") && <HelpText widget={widget} />}
      </div>

      <input
        onChange={onChangeValue}
        className="datepicker"
        type={widget.get("dateType")!}
        required={widget.get("required")}></input>
      <input
        type="hidden"
        className="show-in-review"
        name={getFieldName(widget)}
        value={value}></input>
    </div>
  );
});
