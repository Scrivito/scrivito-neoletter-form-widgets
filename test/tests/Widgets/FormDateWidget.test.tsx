import * as Scrivito from "scrivito";
import { screen } from "@testing-library/react";
import { FormDateWidget } from "../../../src/Widgets/FormDateWidget/FormDateWidgetClass";
import "../../../src/Widgets/FormDateWidget/FormDateWidgetComponent";
import PageRenderer from "../../helpers/pageRenderer";

Scrivito.configure({ tenant: "inMemory" });

const pageRenderer = new PageRenderer();

const widgetProps = {
  title: "Test Date",
  required: true,
  helpText: "This is a test help text",
  dateType: "date",
  customFieldName: "custom_test"
};
describe("FormDateWidget", () => {
  it("displays Mandatory component and HelpText", () => {
    pageRenderer.render({
      body: [new FormDateWidget(widgetProps)]
    });

    const mandatoryText = screen.getByText("*", {
      selector: ".text-mandatory"
    });
    const helpTextElement = document.querySelector("i");
    expect(mandatoryText).toBeInTheDocument();
    expect(helpTextElement).toBeInTheDocument();
  });

  it("renders input fields correctly", () => {
    pageRenderer.render({
      body: [new FormDateWidget(widgetProps)]
    });

    const hiddenInput = document.querySelector('[type="hidden"]')!;
    expect(hiddenInput).toHaveAttribute("name", widgetProps.customFieldName);
    expect(hiddenInput).toHaveClass("show-in-review");
  });

  it("renders correctly", () => {
    const tree = pageRenderer.getAsJSON({
      body: [new FormDateWidget(widgetProps)]
    });
    expect(tree).toMatchSnapshot();
  });
});
