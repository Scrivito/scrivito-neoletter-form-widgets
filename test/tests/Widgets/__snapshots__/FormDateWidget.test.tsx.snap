// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormDateWidget renders correctly 1`] = `
<div>
  <div>
    <div
      className="form-date mb-3"
    >
      <div
        className="date-title"
      >
        <span
          className="text-super"
        >
          Test Date
        </span>
        <div
          className="mandatory-container"
        >
          <span
            className="text-mandatory"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            *
          </span>
        </div>
        <div
          className="helptext-container"
        >
          <i
            className="bi bi-question-circle"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          />
        </div>
      </div>
      <input
        className="datepicker"
        onChange={[Function]}
        required={true}
        type="date"
      />
      <input
        className="show-in-review"
        name="custom_test"
        type="hidden"
        value=""
      />
    </div>
  </div>
</div>
`;
